### PR TITLE
fix(yamleditor): comment indentation shift during scalar replacement

### PR DESCRIPTION
**Pull request by fmudry**
With the new sub-commands that will be added in the PR mentioned below, we would break users comments.

When a scalar replace escalates to a parent block replace, ruamel.yaml preserves absolute column positions for comments while code lines shift to column 0. The subsequent dedent/indent cycle re-indents code lines correctly but over-indents comments by `col` spaces.

Strip `col` characters of leading whitespace from comment-only lines before the dedent/indent cycle so comments stay at their correct relative position within the block. Lines inside block scalars (| / >) are skipped to avoid corrupting literal content that starts with #.

Also updated tests.

### Investigation
I noticed the error while testing the solution of [this PR](https://github.com/konflux-ci/pipeline-migration-tool/pull/137)

<details>
<summary>Investigation</summary>

For testing, I used this file:
```
# Minimal PipelineRun for testing multiple runAfter references to clair-scan
---
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: test-pipeline-three
spec:
  params:
  - name: output-image
    value: quay.io/test/image:latest
  - name: skip-checks
    value: "false"
  - name: build-platforms
    value:
    - linux/x86_64
  pipelineSpec:
    params:
    - name: output-image
      type: string
    - default: "false"
      name: skip-checks
      type: string
    - default:
      - linux/x86_64
      name: build-platforms
      type: array
    tasks:
    # Build the image index
    - name: build-image-index
      params:
      - name: IMAGE
        value: $(params.output-image)
      taskRef:
        params:
        - name: name
          value: build-image-index
        - name: bundle
          value: quay.io/konflux-ci/tekton-catalog/task-build-image-index:0.3.1@sha256:cc75f64deecccb1b59e96ac1182665a5342d79c9e22eebff63d26b0f00a4319c
        - name: kind
          value: task
        resolver: bundles
    # Scan for vulnerabilities
    - matrix:
        params:
        - name: image-platform
          value:
          - $(params.build-platforms)
      name: clair-scan
      params:
      - name: image-digest
        value: $(tasks.build-image-index.results.IMAGE_DIGEST)
      - name: image-url
        value: $(tasks.build-image-index.results.IMAGE_URL)
      runAfter:
      - build-image-index
      taskRef:
        params:
        - name: name
          value: clair-scan
        - name: bundle
          value: quay.io/konflux-ci/tekton-catalog/task-clair-scan:0.3.2@sha256:f5b4415db9ac1fba3e11d993a617e0b275d1f0ed2fc669b12c400ed848c39174
        - name: kind
          value: task
        resolver: bundles
      when:
      - input: $(params.skip-checks)
        operator: in
        values:
        - "false"
    # SAST analysis
    - name: sast-snyk-check-oci-ta
      params:
      - name: image-digest
        value: $(tasks.build-image-index.results.IMAGE_DIGEST)
      - name: image-url
        value: $(tasks.build-image-index.results.IMAGE_URL)
      runAfter:
      - build-image-index
      taskRef:
        params:
        - name: name
          value: sast-snyk-check-oci-ta
        - name: bundle
          value: quay.io/konflux-ci/tekton-catalog/task-sast-snyk-check-oci-ta:0.5@sha256:eba24f5d9f4b18aa71e523b9b3dbcf22982aa4b018824260a090b19dfc9abf6f
        - name: kind
          value: task
        resolver: bundles
      when:
      - input: $(params.skip-checks)
        operator: in
        values:
        - "false"
    # Runs after clair-scan only
    - name: rpms-signature-scan
      params:
      - name: image-digest
        value: $(tasks.build-image-index.results.IMAGE_DIGEST)
      - name: image-url
        value: $(tasks.build-image-index.results.IMAGE_URL)
      runAfter:
      - roxtl-scan
      taskRef:
        params:
        - name: name
          value: rpms-signature-scan
        - name: bundle
          value: quay.io/konflux-ci/tekton-catalog/task-rpms-signature-scan:0.2.1@sha256:ccb77d1bf7627fc6241a59ed42bb6e5707a8682754fe8ae18f2cfdddbcf29275
        - name: kind
          value: task
        resolver: bundles
    # Runs after clair-scan and sast-snyk-check
    - name: deprecated-base-image-check
      params:
      - name: IMAGE_URL
        value: $(tasks.build-image-index.results.IMAGE_URL)
      - name: IMAGE_DIGEST
        value: $(tasks.build-image-index.results.IMAGE_DIGEST)
      runAfter:
      - roxtl-scan
      - sast-snyk-check-oci-ta
      taskRef:
        params:
        - name: name
          value: deprecated-image-check
        - name: bundle
          value: quay.io/konflux-ci/tekton-catalog/task-deprecated-image-check:0.5@sha256:0ccc688a77e9b7b0b8973c132a1e840844137e77f887be4a0bec8893b0776872
        - name: kind
          value: task
        resolver: bundles
      when:
      - input: $(params.skip-checks)
        operator: in
        values:
        - "false"
    # Runs after clair-scan and build-image-index
    - name: push-dockerfile
      params:
      - name: IMAGE
        value: $(tasks.build-image-index.results.IMAGE_URL)
      - name: IMAGE_DIGEST
        value: $(tasks.build-image-index.results.IMAGE_DIGEST)
      runAfter:
      - roxtl-scan
      - build-image-index
      taskRef:
        params:
        - name: name
          value: push-dockerfile-oci-ta
        - name: bundle
          value: quay.io/konflux-ci/tekton-catalog/task-push-dockerfile-oci-ta:0.3.1@sha256:e19d134048f05dc7c337914b68590699240b92194979bfcc9e1ef4bdf6adb4c4
        - name: kind
          value: task
        resolver: bundles
```

And after running:
```bash
pmt modify -f pipeline-three.yaml task clair-scan rename roxtl-scan --task-ref-name task-roxtl-scan
```
I got (plus 4 spaces):
```diff
-   # SAST analysis
    - name: sast-snyk-check-oci-ta
+      # SAST analysis
    - name: sast-snyk-check-oci-ta
```
Which was in a block which shouldn't be touched.

</details>

### CODE REVIEW

File: src/pipeline_migration/yamleditor.py [hunk 1]
```diff
@@ -29,6 +29,29 @@
...
+    """
+    if strip_spaces <= 0:
+        return yaml_output
+    lines = yaml_output.split("\n")
+    result = []
+    for line in lines:
+        stripped = line.lstrip()
+        if stripped.startswith("#"):
+            indent = len(line) - len(stripped)
+            result.append(" " * max(0, indent - strip_spaces) + stripped)
```
[the remark this answers is not in the document]
Comment (fmudry) [reply src/pipeline_migration/yamleditor.py]: qodo was right, fixed.

File: tests/test_yamleditor.py [hunk 2]
```diff
@@ -1361,6 +1361,115 @@ def test_replace_empty_commented_section(self, comments_yaml_file):
...
+                  taskRef:
+                    resolver: bundles
+            """)
+        yaml_file = create_yaml_file(content)
+        editor = EditYAMLEntry(yaml_file)
+
+        editor.replace(["spec", "tasks", 0, "name"], "roxtl-scan")
+
+        result = read_file_content(yaml_file)
+        assert "    # SAST analysis\n    - name: sast-snyk-check" in result
```
Comment (chmeliik) [inline tests/test_yamleditor.py]: It's not clear to me how the asserts in these tests check the proper behavior. In fact, I think they don't, since all the new tests pass in main as well

**Review by chmeliik — COMMENTED** [review]
I'm afraid that handling the YAML content as a string opens us up to many corner cases.

For example, here's a file where pmt mistakenly identifies a heredoc and over-indents the comment anyway:

<details>
<summary>03-false-multiline-marker.yaml</summary>

```yaml
spec:
  pipelineSpec:
    tasks:
    - name: foo
      wouldYouBelieveThisIsAString: :>3
      # some comment
      taskRef:
        params:
        - name: name
          value: foo
```

```bash
$ pmt modify -f 03-false-multiline-marker.yaml generic replace '["spec", "pipelineSpec", "tasks", 0, "name"]' "bar"

$ git diff 03-false-multiline-marker.yaml
```

```diff
 spec:
   pipelineSpec:
     tasks:
-    - name: foo
+    - name: bar
       wouldYouBelieveThisIsAString: :>3
-      # some comment
+          # some comment
       taskRef:
         params:
         - name: name
```

</details>

The current approach also doesn't work for inline comments (still over-indents):

<details>
<summary>01-inline-comment.yaml</summary>

```yaml
spec:
  pipelineSpec:
    tasks:
    - name: foo
      taskRef:  # some comment
        params:
        - name: name
          value: foo
```

```bash
$ pmt modify -f 01-inline-comment.yaml generic replace '["spec", "pipelineSpec", "tasks", 0, "name"]' "bar"

$ git diff 01-inline-comment.yaml
```

```diff
 spec:
   pipelineSpec:
     tasks:
-    - name: foo
-      taskRef:  # some comment
+    - name: bar
+      taskRef:      # some comment
         params:
         - name: name
           value: foo
```

</details>

Or for files that use 2-space indentation for lists (under-indents, but this should be more easily fixable):

<details>
<summary>02-twospace-indent.yaml</summary>

```yaml
spec:
  pipelineSpec:
    tasks:
      - name: foo
        # some comment
        taskRef:
          params:
            - name: name
              value: foo
```

```bash
$ pmt modify -f 02-twospace-indent.yaml generic replace '["spec", "pipelineSpec", "tasks", 0, "name"]' "bar"

$ git diff 02-twospace-indent.yaml
```

```diff
 spec:
   pipelineSpec:
     tasks:
-      - name: foo
-        # some comment
+      - name: bar
+      # some comment
         taskRef:
           params:
             - name: name
```

</details>

An alternative approach might be to fix the column numbers and string values for the [`CommentToken`](https://github.com/commx/ruamel-yaml/blob/b6b5921256e7b34c325be5af027cc7f7d68d2f14/tokens.py#L334) objects rather than editing the raw YAML string. I've played with it a bit and the ruamel.yaml API is god-awful, but it seems doable. Every object that inherits from `CommentedBase` has a `ca` attribute of type `Comment`. I used the [`Comment.__contains__`](https://github.com/commx/ruamel-yaml/blob/b6b5921256e7b34c325be5af027cc7f7d68d2f14/comments.py#L177) method as inspiration for how to actually get the damn comments.

File: tests/test_yamleditor.py (hunk 2 above)
Comment (fmudry) [reply tests/test_yamleditor.py]: Good catch, I need to be more careful with the `in` operator... :facepalm:

**Comment by fmudry** [comment]
@chmeliik  Thanks for your great comment!

I have made changes that move away from raw yaml strings manipulation. Seems to be more reliable manipulating `CommentToken`. :smile:

I have fixed the tests and added a few more. There are still two tests which pass on main (`preserves_block_scalar_hash_lines` and `with_block_scalar_inline_comment`), but I would still keep them as "guards" for the future. Especially the second one, since ruamel stores some tokens as just `str`, which we want to skip.

Let me know if you see room for improvement :heart_hands:

File: src/pipeline_migration/yamleditor.py [hunk 3]
```diff
@@ -29,6 +29,73 @@
...
+    if token is None or isinstance(token, str):  # ruamel stores some slots as plain str
+        return
+    if hasattr(token, "start_mark") and token.start_mark is not None:
+        token.start_mark.column = max(0, token.start_mark.column - offset)
+    value = token.value
+    if isinstance(value, str) and "\n" in value:  # multi-line: fix embedded indentation
+        lines = value.split("\n")
+        adjusted: list[str] = []
+        for line in lines:
+            if line.lstrip().startswith("#"):  # skip blank separator lines
```
Comment (chmeliik) [inline src/pipeline_migration/yamleditor.py:48]: nitpick: I don't think we have to skip blank separator lines, we can (probably should) strip leading whitespace from all lines

File: src/pipeline_migration/yamleditor.py [hunk 4]
```diff
@@ -391,9 +458,23 @@ def find_next_sibling_line(node, index) -> int | None:
         return EOF
 
     def _gen_yaml_str(self, data: Any, col: int, seq_block: bool = False) -> str:
-        """Generate an indented YAML string for the given data at the specified column."""
+        """Generate an indented YAML string for the given data at the specified column.
+
+        Comment columns are pre-adjusted so the dedent/indent cycle
+        preserves their alignment relative to surrounding code.
```
Comment (chmeliik) [inline src/pipeline_migration/yamleditor.py:464]: I think mentioning the dedent/indent indent cycle adds unnecessary confusion. It's the ruamel.yaml round-tripping itself that breaks comment positioning (if you `load()` a parent object and `dump()` a child path from that object).

The dedent/indent doesn't play much of a part in it. It's a workaround for the bug where ruamel.yaml indents the entire dumped string if the root object is a list. We need to take that bug into account when calculating the offset, but it's not the main reason why we need to fix comment positioning

File: src/pipeline_migration/yamleditor.py [hunk 5]
```diff
@@ -391,9 +458,23 @@ def find_next_sibling_line(node, index) -> int | None:
...
+            data = copy.deepcopy(data)  # avoid mutating caller's comment metadata
         if seq_block:
             data = [data]
+        # With a non-zero block sequence offset the dump indents `- ` by that
+        # offset, which textwrap.dedent later strips from *all* lines including
+        # comments.  Reduce the comment adjustment accordingly so the net result
+        # (adjust → dump → dedent → indent) lands comments at their original column.
+        comment_offset = col
+        if seq_block and self.style and self.style.indentation.is_consistent:
+            comment_offset = max(0, col - self.style.indentation.levels[0])  # dedent eats this
```
Comment (chmeliik) [inline src/pipeline_migration/yamleditor.py:476]: I'd suggest to `create_yaml_obj()` first and then reduce the offset by `yaml.sequence_dash_offset` to avoid having to replicate part of the indentation detection logic

(`yaml.sequence_dash_offset` is a value already calculated based on `self.style.indentation`)

File: tests/test_yamleditor.py [hunk 6]
```diff
@@ -1361,6 +1361,231 @@ def test_replace_empty_commented_section(self, comments_yaml_file):
...
+        lines = result.splitlines()
+        comment_line = next(ln for ln in lines if "# some comment" in ln)
+        assert comment_line == "        # some comment"
+
+    def test_replace_scalar_with_block_scalar_inline_comment(self, create_yaml_file):
+        """Inline comment on a block scalar indicator (|  # hello) is preserved.
+
+        ruamel.yaml stores these comments as plain str objects in ca.items,
+        not as CommentToken.  Without a guard for str, _adjust_comment_token
+        crashes with AttributeError on token.value.
```
Comment (chmeliik) [inline tests/test_yamleditor.py:1557]: 🤯 What a well-designed library. This case even crashes ruamel.yaml itself, but even earlier:

```python
import ruamel.yaml

raw_doc = """
what: |  # the hell
  is this?
"""

yaml = ruamel.yaml.YAML()

doc = yaml.load(raw_doc)
print("hell" in doc.ca)
```

```text
Traceback (most recent call last):
  File "/tmp/x.py", line 11, in <module>
    print("hell" in doc.ca)
          ^^^^^^^^^^^^^^^^
  File "/home/acmiel/RedHat/Stonesoup/pipeline-migration-tool/.venv/lib64/python3.14/site-packages/ruamel/yaml/comments.py", line 190, in __contains__
    if c and x in c.value:
                  ^^^^^^^
AttributeError: 'list' object has no attribute 'value'
```

It even fails to handle ca.items values that are lists, let alone plain strings within those lists.

Good find 🌮

**Review by chmeliik — COMMENTED** [review]
Nice job! Some small suggestions, but this looks solid 🎉

File: tests/test_yamleditor.py (hunk 6 above)
Comment (fmudry) [reply tests/test_yamleditor.py:1557]: As Gandalf once said: For even the very wise cannot see all ends :military_helmet:

**Comment by fmudry** [comment]
Thanks! Updated the code based on your suggestions.

**Review by chmeliik — APPROVED** [review]

File: tests/test_yamleditor.py (hunk 6 above)
Comment (chmeliik) [reply tests/test_yamleditor.py:1557]: image
